Так как сказали ускорится и до завтра сделать оставив некоторые недочеты некая предзащита нам надо ускорится и игратся в тестововй ветке уже нет времени увы 

## Необходимые библиотеки

In [23]:
import pandas as pd
import numpy as np
import ast
import re
import torch
from collections import Counter
# from iterstrat.ml_stratifiers import MultilabelStratifiedKFold Не совсем подходит для моих представлений и целей
from skmultilearn.model_selection import iterative_train_test_split
from sklearn.metrics import classification_report
from transformers import BertTokenizer, BertForSequenceClassification, Trainer, TrainingArguments
from transformers import RobertaTokenizer, RobertaForSequenceClassification, Trainer, TrainingArguments
from transformers import BertTokenizer, BertForSequenceClassification, Trainer, TrainingArguments
from torch.utils.data import Dataset
from sklearn.metrics import classification_report, f1_score, hamming_loss, accuracy_score, roc_auc_score

## Просмотр изначальных данных

In [24]:
dt = pd.read_csv("original.csv")
dt

id  rating                                            comment  \
0     2945792       5                                            спасибо   
1     3234340       5                                           спасибо!   
2     3380332       5                                           Отлично    
3     3381812       5       Благодарю за оперативное решение проблемы !    
4     3461991       5              Прекрасный специалист! Побольше таких   
...       ...     ...                                                ...   
2368  3507812       5                                  долго, но приятно   
2369  3478532       5  16.03.2024 г ребята установили кондиционер. ра...   
2370  3296403       5          два дня мусоропровод. сегодня с утра тоже   
2371  3165822       5  отрегулировать дверь замки . для должного откр...   
2372  2991660       5                                 всё сделано хорошо   

      Нравится скорость отработки заявок  Нравится качество выполнения заявки  \
0                                    NaN                                  NaN   
1                                    NaN                                  NaN   
2                                    NaN                                  NaN   
3                                    NaN                                  NaN   
4                                    NaN                                  NaN   
...                                  ...                                  ...   
2368                                 NaN                                  NaN   
2369                                 NaN                                  NaN   
2370                                 NaN                                  NaN   
2371                                 NaN                                  NaN   
2372                                 NaN                                  NaN   

      Нравится качество работы сотрудников  Понравилось выполнение заявки  \
0                                      NaN                            NaN   
1                                      NaN                            NaN   
2                                      NaN                            NaN   
3                                      NaN                            NaN   
4                                      NaN                            NaN   
...                                    ...                            ...   
2368                                   NaN                            NaN   
2369                                   NaN                            NaN   
2370                                   NaN                            NaN   
2371                                   NaN                            NaN   
2372                                   NaN                            NaN   

      Вопрос решен  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN  
...            ...  
2368           NaN  
2369           NaN  
2370           NaN  
2371           NaN  
2372           NaN  

[2373 rows x 8 columns]

#### У нас 2373 неразмеченных данных и 5 категорий для разметки


## Уже размеченные данные

#### Я разметил наши данные через label-studio

In [25]:
df = pd.read_csv("marked_up.csv")
df

annotation_id  annotator  \
0                16          2   
1                17          2   
2                18          2   
3                19          2   
4                20          2   
...             ...        ...   
2363           2379          2   
2364           2380          2   
2365           2381          2   
2366           2382          2   
2367           2383          2   

                                                comment  \
0                                               спасибо   
1                                              спасибо!   
2                                              Отлично    
3          Благодарю за оперативное решение проблемы !    
4                 Прекрасный специалист! Побольше таких   
...                                                 ...   
2363                                  долго, но приятно   
2364  16.03.2024 г ребята установили кондиционер. ра...   
2365          два дня мусоропровод. сегодня с утра тоже   
2366  отрегулировать дверь замки . для должного откр...   
2367                                 всё сделано хорошо   

                       created_at       id  lead_time  rating  \
0     2025-05-08T12:32:06.806278Z  2945792     49.604       5   
1     2025-05-08T12:44:25.556719Z  3234340      2.058       5   
2     2025-05-08T12:44:46.429049Z  3380332     19.677       5   
3     2025-05-08T12:45:09.100228Z  3381812     21.395       5   
4     2025-05-08T12:45:23.515860Z  3461991     13.272       5   
...                           ...      ...        ...     ...   
2363  2025-05-10T20:12:27.250167Z  3507812      2.124       5   
2364  2025-05-10T20:12:37.420217Z  3478532      9.359       5   
2365  2025-05-10T20:12:47.220202Z  3296403      8.697       5   
2366  2025-05-10T20:12:52.522679Z  3165822      4.581       5   
2367  2025-05-10T20:12:55.135450Z  2991660      1.833       5   

                                              sentiment  \
0                                          Вопрос решен   
1                                          Вопрос решен   
2     {"choices":["Понравилось выполнение заявки","В...   
3     {"choices":["Нравится скорость отработки заяво...   
4                  Нравится качество работы сотрудников   
...                                                 ...   
2363                                       Вопрос решен   
2364  {"choices":["Нравится скорость отработки заяво...   
2365                                       Вопрос решен   
2366                       ни к одному из перечисленных   
2367                                       Вопрос решен   

                       updated_at  Вопрос решен  \
0     2025-05-08T12:44:22.535929Z           NaN   
1     2025-05-08T12:44:25.556719Z           NaN   
2     2025-05-08T12:44:46.429049Z           NaN   
3     2025-05-08T12:45:09.100228Z           NaN   
4     2025-05-08T12:45:23.515860Z           NaN   
...                           ...           ...   
2363  2025-05-10T20:12:27.251168Z           NaN   
2364  2025-05-10T20:12:37.420217Z           NaN   
2365  2025-05-10T20:12:47.220202Z           NaN   
2366  2025-05-10T20:12:52.522679Z           NaN   
2367  2025-05-10T20:12:55.135450Z           NaN   

      Нравится качество выполнения заявки  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                     NaN   
...                                   ...   
2363                                  NaN   
2364                                  NaN   
2365                                  NaN   
2366                                  NaN   
2367                                  NaN   

      Нравится качество работы сотрудников  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4             

#### это размеченные наши данные гдето потерялось 5 коментариев :(

## Анализ и доработка данных

#### Ну тут очевидно что нам не нужно большинство колонок

In [26]:
df = df.drop(['id' ,'created_at', 'updated_at', 'annotation_id', 'annotator', 'lead_time'], axis=1)
# Смотрим или мы все верно удалили ничего не забыли и ничего лишнего не зацепили(так у нас id не удалился а почему)авхахвп нето вывел ладно
df

comment  rating  \
0                                               спасибо       5   
1                                              спасибо!       5   
2                                              Отлично        5   
3          Благодарю за оперативное решение проблемы !        5   
4                 Прекрасный специалист! Побольше таких       5   
...                                                 ...     ...   
2363                                  долго, но приятно       5   
2364  16.03.2024 г ребята установили кондиционер. ра...       5   
2365          два дня мусоропровод. сегодня с утра тоже       5   
2366  отрегулировать дверь замки . для должного откр...       5   
2367                                 всё сделано хорошо       5   

                                              sentiment  Вопрос решен  \
0                                          Вопрос решен           NaN   
1                                          Вопрос решен           NaN   
2     {"choices":["Понравилось выполнение заявки","В...           NaN   
3     {"choices":["Нравится скорость отработки заяво...           NaN   
4                  Нравится качество работы сотрудников           NaN   
...                                                 ...           ...   
2363                                       Вопрос решен           NaN   
2364  {"choices":["Нравится скорость отработки заяво...           NaN   
2365                                       Вопрос решен           NaN   
2366                       ни к одному из перечисленных           NaN   
2367                                       Вопрос решен           NaN   

      Нравится качество выполнения заявки  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                     NaN   
...                                   ...   
2363                                  NaN   
2364                                  NaN   
2365                                  NaN   
2366                                  NaN   
2367                                  NaN   

      Нравится качество работы сотрудников  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                      NaN   
...                                    ...   
2363                                   NaN   
2364                                   NaN   
2365                                   NaN   
2366                                   NaN   
2367                                   NaN   

      Нравится скорость отработки заявок  Понравилось выполнение заявки  
0                                    NaN                            NaN  
1                                    NaN                            NaN  
2                                    NaN                            NaN  
3                                    NaN                            NaN  
4                                    NaN                            NaN  
...                                  ...                            ...  
2363                                 NaN                            NaN  
2364                                 NaN                            NaN  
2365                                 NaN                            NaN  
2366                                 NaN                            NaN  
2367                                 NaN                            NaN  

[2368 rows x 8 columns]

#### Далее надо добавить еще 1 колонку так ка она добавилась в ходе разметки из-за не решонных вопросовы или негатиувных коментариев

In [27]:
df['ни к одному из перечисленных'] = 0
df

comment  rating  \
0                                               спасибо       5   
1                                              спасибо!       5   
2                                              Отлично        5   
3          Благодарю за оперативное решение проблемы !        5   
4                 Прекрасный специалист! Побольше таких       5   
...                                                 ...     ...   
2363                                  долго, но приятно       5   
2364  16.03.2024 г ребята установили кондиционер. ра...       5   
2365          два дня мусоропровод. сегодня с утра тоже       5   
2366  отрегулировать дверь замки . для должного откр...       5   
2367                                 всё сделано хорошо       5   

                                              sentiment  Вопрос решен  \
0                                          Вопрос решен           NaN   
1                                          Вопрос решен           NaN   
2     {"choices":["Понравилось выполнение заявки","В...           NaN   
3     {"choices":["Нравится скорость отработки заяво...           NaN   
4                  Нравится качество работы сотрудников           NaN   
...                                                 ...           ...   
2363                                       Вопрос решен           NaN   
2364  {"choices":["Нравится скорость отработки заяво...           NaN   
2365                                       Вопрос решен           NaN   
2366                       ни к одному из перечисленных           NaN   
2367                                       Вопрос решен           NaN   

      Нравится качество выполнения заявки  \
0                                     NaN   
1                                     NaN   
2                                     NaN   
3                                     NaN   
4                                     NaN   
...                                   ...   
2363                                  NaN   
2364                                  NaN   
2365                                  NaN   
2366                                  NaN   
2367                                  NaN   

      Нравится качество работы сотрудников  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                      NaN   
...                                    ...   
2363                                   NaN   
2364                                   NaN   
2365                                   NaN   
2366                                   NaN   
2367                                   NaN   

      Нравится скорость отработки заявок  Понравилось выполнение заявки  \
0                                    NaN                            NaN   
1                                    NaN                            NaN   
2                                    NaN                            NaN   
3                                    NaN                            NaN   
4                                    NaN                            NaN   
...                                  ...                            ...   
2363                                 NaN                            NaN   
2364                                 NaN                            NaN   
2365                                 NaN                            NaN   
2366                                 NaN                            NaN   
2367                                 NaN                            NaN   

      ни к одному из перечисленных  
0                                0  
1                                0  
2                                0  
3                                0  
4                                0  
...                            ...  
2363                             0  
2364                             0  
2365                             0  
2

#### Мы добавили колонку теперь перенесем разметку по классам из колонки Sentiment и удалим колонку

In [28]:

stolbiki = [
    'Вопрос решен',
    'Нравится качество выполнения заявки',
    'Нравится качество работы сотрудников',
    'Нравится скорость отработки заявок',
    'Понравилось выполнение заявки',
    'ни к одному из перечисленных'
]

for idx, row in df.iterrows():
    try:
        sentiment = row['sentiment']
        if isinstance(sentiment, str) and '"choices"' in sentiment:
            data = ast.literal_eval(sentiment)
            for label in data.get("choices", []):
                if label in stolbiki:
                    df.at[idx, label] = 1
        elif sentiment in stolbiki:
            df.at[idx, sentiment] = 1
    except Exception as e:
        print(f"Ошибка в строке {idx}: {e}")

#### Посмотим или перенеслось и переименуем колонку 'ни к одному из перечисленных' на более каректное имя 

In [29]:
df = df.rename(columns={'ни к одному из перечисленных': 'Проблемы'})
colum = ['Вопрос решен',
    'Нравится качество выполнения заявки',
    'Нравится качество работы сотрудников',
    'Нравится скорость отработки заявок',
    'Понравилось выполнение заявки',
    'Проблемы']
df[colum] = df[colum].fillna(0)
summa = df[colum].sum()
summa
df = df.drop(['sentiment'], axis=1)


#### Мы посмотрели количество данных  в каждом классов и заполнили пустые значения 0

#### Теперь посмотрик какие слова встречаются чаще в каждом из классов

In [30]:
# def propusk():
#     print("\n"*2)
# для 'Вопрос решен'
klass_bopr = df[df['Вопрос решен'] != 0]
klass_bopr = klass_bopr.drop(['Нравится качество выполнения заявки',
    'Нравится качество работы сотрудников',
    'Нравится скорость отработки заявок',
    'Понравилось выполнение заявки',
    'Проблемы'], axis=1)
klass_bopr['comment'] = klass_bopr['comment'].str.lower()
klass_bopr['comment'] = klass_bopr['comment'].apply(lambda x: re.sub(r"[^а-яёa-z\s]", "", x))
all_words = []
for comment in klass_bopr['comment']:
    words = comment.split()
    all_words.extend(words)
word_counts = Counter(all_words)
top_words = word_counts.most_common(10)
print("Топ-10 самых частых слов в классе 'Вопрос решен'")
for word, count in top_words:
    print(f"{word}: {count} раз")


Топ-10 самых частых слов в классе 'Вопрос решен'
спасибо: 1319 раз
за: 380 раз
быстро: 354 раз
и: 340 раз
большое: 271 раз
все: 270 раз
оперативно: 258 раз
очень: 170 раз
оперативность: 134 раз
в: 131 раз


In [31]:
# для 'Нравится качество выполнения заявки'

klass_bopr = df[df['Нравится качество выполнения заявки'] != 0]

klass_bopr = klass_bopr.drop(['Вопрос решен',
    'Нравится качество работы сотрудников',
    'Нравится скорость отработки заявок',
    'Понравилось выполнение заявки',
    'Проблемы'], axis=1)

klass_bopr['comment'] = klass_bopr['comment'].str.lower()
klass_bopr['comment'] = klass_bopr['comment'].apply(lambda x: re.sub(r"[^а-яёa-z\s]", "", x))
all_words = []
for comment in klass_bopr['comment']:
    words = comment.split()
    all_words.extend(words)
word_counts = Counter(all_words)
top_words = word_counts.most_common(10)
print("Топ-10 самых частых слов в классе 'Нравится качество выполнения заявки'")
for word, count in top_words:
    print(f"{word}: {count} раз")

Топ-10 самых частых слов в классе 'Нравится качество выполнения заявки'
и: 100 раз
качественно: 85 раз
быстро: 81 раз
спасибо: 78 раз
все: 38 раз
за: 29 раз
оперативно: 26 раз
очень: 23 раз
большое: 23 раз
работу: 20 раз


In [32]:
# для Нравится качество работы сотрудников'

klass_bopr = df[df['Нравится качество работы сотрудников'] != 0]

klass_bopr = klass_bopr.drop(['Вопрос решен',
    'Нравится качество выполнения заявки',
    'Нравится скорость отработки заявок',
    'Понравилось выполнение заявки',
    'Проблемы'], axis=1)

klass_bopr['comment'] = klass_bopr['comment'].str.lower()
klass_bopr['comment'] = klass_bopr['comment'].apply(lambda x: re.sub(r"[^а-яёa-z\s]", "", x))
all_words = []
for comment in klass_bopr['comment']:
    words = comment.split()
    all_words.extend(words)
word_counts = Counter(all_words)
top_words = word_counts.most_common(10)
print("Топ-10 самых частых слов в классе 'Нравится качество работы сотрудников'")
for word, count in top_words:
    print(f"{word}: {count} раз")

Топ-10 самых частых слов в классе 'Нравится качество работы сотрудников'
спасибо: 233 раз
и: 136 раз
все: 96 раз
быстро: 88 раз
мастер: 84 раз
за: 82 раз
очень: 64 раз
большое: 61 раз
оперативно: 49 раз
в: 49 раз


In [33]:
# для 'Нравится скорость отработки заявок'
klass_bopr = df[df['Нравится скорость отработки заявок'] != 0]

klass_bopr = klass_bopr.drop(['Вопрос решен',

'Нравится качество выполнения заявки',
    'Нравится качество работы сотрудников',
    'Понравилось выполнение заявки',
    'Проблемы'], axis=1)

klass_bopr['comment'] = klass_bopr['comment'].str.lower()
klass_bopr['comment'] = klass_bopr['comment'].apply(lambda x: re.sub(r"[^а-яёa-z\s]", "", x))
all_words = []
for comment in klass_bopr['comment']:
    words = comment.split()
    all_words.extend(words)
word_counts = Counter(all_words)
top_words = word_counts.most_common(10)
print("Топ-10 самых частых слов в классе 'Нравится скорость отработки заявок'")
for word, count in top_words:
    print(f"{word}: {count} раз")

Топ-10 самых частых слов в классе 'Нравится скорость отработки заявок'
спасибо: 464 раз
быстро: 336 раз
оперативно: 253 раз
за: 251 раз
и: 224 раз
оперативность: 135 раз
очень: 119 раз
все: 119 раз
большое: 92 раз
качественно: 76 раз


In [34]:
# для класса 'Понравилось выполнение заявки'
klass_bopr = df[df['Понравилось выполнение заявки'] != 0]

klass_bopr = klass_bopr.drop(['Вопрос решен',

'Нравится качество выполнения заявки',
    'Нравится качество работы сотрудников',
    'Нравится скорость отработки заявок',
    'Проблемы']
, axis=1)

klass_bopr['comment'] = klass_bopr['comment'].str.lower()
klass_bopr['comment'] = klass_bopr['comment'].apply(lambda x: re.sub(r"[^а-яёa-z\s]", "", x))
all_words = []
for comment in klass_bopr['comment']:
    words = comment.split()
    all_words.extend(words)
word_counts = Counter(all_words)
top_words = word_counts.most_common(10)
print("Топ-10 самых частых слов в классе 'Понравилось выполнение заявки'")
for word, count in top_words:
    print(f"{word}: {count} раз")

Топ-10 самых частых слов в классе 'Понравилось выполнение заявки'
спасибо: 206 раз
и: 114 раз
отлично: 103 раз
все: 102 раз
быстро: 85 раз
за: 65 раз
большое: 52 раз
всё: 51 раз
очень: 43 раз
оперативно: 41 раз


In [35]:
# для класса 'Проблемы'
klass_bopr = df[df['Проблемы'] != 0]

klass_bopr = klass_bopr.drop(['Вопрос решен',

'Нравится качество выполнения заявки',
    'Нравится качество работы сотрудников',
    'Нравится скорость отработки заявок',
    'Понравилось выполнение заявки']
, axis=1)

klass_bopr['comment'] = klass_bopr['comment'].str.lower()
klass_bopr['comment'] = klass_bopr['comment'].apply(lambda x: re.sub(r"[^а-яёa-z\s]", "", x))
all_words = []
for comment in klass_bopr['comment']:
    words = comment.split()
    all_words.extend(words)
word_counts = Counter(all_words)
top_words = word_counts.most_common(10)
print("Топ-10 самых частых слов в классе 'Проблемы'")
for word, count in top_words:
    print(f"{word}: {count} раз")

Топ-10 самых частых слов в классе 'Проблемы'
не: 143 раз
в: 94 раз
и: 76 раз
на: 49 раз
что: 38 раз
за: 32 раз
по: 26 раз
с: 26 раз
так: 25 раз
нет: 24 раз


##### суде по всему людям при плохо выполненной работе или не решонном вопросе любят использовать предлоги, союзы и местоимения 

## Очистка данных


#### Удаление стоп-слов и лемматизация мы делать не будем так как у нас они могут нести смысловую нагрузку 

#### удалим лишние пробелы 

In [36]:

df['comment'] = df['comment'].str.replace(r'\s+', ' ', regex=True)
df

comment  rating  Вопрос решен  \
0                                               спасибо       5           1.0   
1                                              спасибо!       5           1.0   
2                                              Отлично        5           1.0   
3          Благодарю за оперативное решение проблемы !        5           1.0   
4                 Прекрасный специалист! Побольше таких       5           0.0   
...                                                 ...     ...           ...   
2363                                  долго, но приятно       5           1.0   
2364  16.03.2024 г ребята установили кондиционер. ра...       5           1.0   
2365          два дня мусоропровод. сегодня с утра тоже       5           1.0   
2366  отрегулировать дверь замки . для должного откр...       5           0.0   
2367                                 всё сделано хорошо       5           1.0   

      Нравится качество выполнения заявки  \
0                                     0.0   
1                                     0.0   
2                                     0.0   
3                                     0.0   
4                                     0.0   
...                                   ...   
2363                                  0.0   
2364                                  1.0   
2365                                  0.0   
2366                                  0.0   
2367                                  0.0   

      Нравится качество работы сотрудников  \
0                                      0.0   
1                                      0.0   
2                                      0.0   
3                                      0.0   
4                                      1.0   
...                                    ...   
2363                                   0.0   
2364                                   1.0   
2365                                   0.0   
2366                                   0.0   
2367                                   0.0   

      Нравится скорость отработки заявок  Понравилось выполнение заявки  \
0                                    0.0                            0.0   
1                                    0.0                            0.0   
2                                    0.0                            1.0   
3                                    1.0                            0.0   
4                                    0.0                            0.0   
...                                  ...                            ...   
2363                                 0.0                            0.0   
2364                                 1.0                            1.0   
2365                                 0.0                            0.0   
2366                                 0.0                            0.0   
2367                                 0.0                            0.0   

      Проблемы  
0            0  
1            0  
2            0  
3            0  
4            0  
...        ...  
2363         0  
2364         1  
2365         0  
2366         1  
2367         0  

[2368 rows x 8 columns]

## Stratified( или просто процентное уравнивание классов)


#### Буду использовать scikit-multilearn так как насколько я понял при делении клсаов через iterstrat.ml_stratifiers не дает возможности делить данные 80 на 20 и мне посоветовали scikit-multilearn

In [37]:
cm = df["comment"].tolist()
lb = df[['Вопрос решен',
    'Нравится качество выполнения заявки',
    'Нравится качество работы сотрудников',
    'Нравится скорость отработки заявок',
    'Понравилось выполнение заявки',
    'Проблемы']].values

# mlskf = MultilabelStratifiedKFold(n_splits=2, shuffle=True, random_state=42)
# for train_idx, test_idx in mlskf.split(cm, lb):
#     X_train, X_test = [cm[i] for i in train_idx], [cm[i] for i in test_idx]
#     y_train, y_test = lb[train_idx], lb[test_idx]

texts_array = np.array(cm).reshape(-1, 1)

X_train, y_train, X_test, y_test = iterative_train_test_split(
    texts_array,
    lb,
    test_size=0.2
)


X_train = X_train.reshape(-1).tolist()
X_test = X_test.reshape(-1).tolist()

train_label_percent = y_train.sum(axis=0) / y_train.sum() * 100
test_label_percent = y_test.sum(axis=0) / y_test.sum() * 100

print("\nВариант: Процент меток класса")
print("Train:", np.round(train_label_percent, 2), "%")
print("Test:", np.round(test_label_percent, 2), "%")


Вариант: Процент меток класса
Train: [50.22  3.66  9.84 19.79 10.31  6.17] %
Test: [50.18  3.66  9.8  19.83 10.39  6.14] %


#### ну в процентном соотношении схоже

## Модельки

In [38]:
print(f"CUDA доступен: {torch.cuda.is_available()}")
print(f"Количество GPU: {torch.cuda.device_count()}")
print(f"Название GPU: {torch.cuda.get_device_name(0)}")

CUDA доступен: True
Количество GPU: 1
Название GPU: NVIDIA GeForce RTX 4070


RuBERT

In [39]:


MODEL_NAME = "DeepPavlov/rubert-base-cased"


tokenizer = BertTokenizer.from_pretrained(MODEL_NAME)


class CustomDataset(Dataset):
    def __init__(self, texts, labels):
        self.encodings = tokenizer(
            texts, padding=True, truncation=True, max_length=512, return_tensors="pt"
        )
        self.labels = torch.tensor(labels, dtype=torch.float)

    def __getitem__(self, idx):
        return {
            'input_ids': self.encodings['input_ids'][idx],
            'attention_mask': self.encodings['attention_mask'][idx],
            'labels': self.labels[idx]
        }

    def __len__(self):
        return len(self.labels)

TEXTS_TRAIN = X_train
Y_TRAIN = y_train
TEXTS_TEST = X_test
Y_TEST = y_test


train_dataset = CustomDataset(TEXTS_TRAIN, Y_TRAIN)
test_dataset = CustomDataset(TEXTS_TEST, Y_TEST)


model = BertForSequenceClassification.from_pretrained(
    MODEL_NAME,
    num_labels=Y_TRAIN.shape[1],
    problem_type="multi_label_classification"
)


training_args = TrainingArguments(
    output_dir='./rubert_results',
    per_device_train_batch_size=7,               
    gradient_accumulation_steps=2,                
    num_train_epochs=4,
    eval_strategy="epoch",
    save_strategy="epoch",
    logging_dir='./logs_rubert',
    logging_steps=50,
    fp16=True,                                  
    dataloader_num_workers=0,                     
    load_best_model_at_end=True,
    metric_for_best_model="eval_loss"
)


trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=test_dataset
)


trainer.train()


predictions = trainer.predict(test_dataset)
y_pred_logits = predictions.predictions
y_pred = (torch.sigmoid(torch.tensor(y_pred_logits)) > 0.5).int().numpy()


print("\nClassification Report:")
print(classification_report(Y_TEST, y_pred, digits=4))

print("\nДополнительные метрики:")
print(f"Micro F1-score:     {f1_score(Y_TEST, y_pred, average='micro'):.4f}")
print(f"Macro F1-score:     {f1_score(Y_TEST, y_pred, average='macro'):.4f}")
print(f"Weighted F1-score:  {f1_score(Y_TEST, y_pred, average='weighted'):.4f}")
print(f"Hamming loss:       {hamming_loss(Y_TEST, y_pred):.4f}")
print(f"Subset accuracy:    {accuracy_score(Y_TEST, y_pred):.4f} (все метки совпали)")



y_pred_prob = torch.sigmoid(torch.tensor(y_pred_logits)).numpy()

roc_auc = roc_auc_score(Y_TEST, y_pred_prob, average='macro')
print(f"ROC AUC (macro): {roc_auc:.4f}")



Some weights of BertForSequenceClassification were not initialized from the model checkpoint at DeepPavlov/rubert-base-cased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.



Classification Report:
              precision    recall  f1-score   support

           0     0.9545    0.9882    0.9711       425
           1     0.8667    0.8387    0.8525        31
           2     0.9036    0.9036    0.9036        83
           3     0.9412    0.9524    0.9467       168
           4     0.6310    0.6023    0.6163        88
           5     0.8750    0.6731    0.7609        52

   micro avg     0.9079    0.9079    0.9079       847
   macro avg     0.8620    0.8264    0.8418       847
weighted avg     0.9052    0.9079    0.9055       847
 samples avg     0.9172    0.9205    0.9062       847


Дополнительные метрики:
Micro F1-score:     0.9079
Macro F1-score:     0.8418
Weighted F1-score:  0.9055
Hamming loss:       0.0549
Subset accuracy:    0.7426 (все метки совпали)
ROC AUC (macro): 0.9491


c:\Users\josha\OneDrive\Documents\GitHub\4_intensiv_4\.venv\Lib\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in samples with no true labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [ ]:

model.save_pretrained("./saved_model0")
tokenizer.save_pretrained("./saved_model0")

In [40]:
class CustomDataset(Dataset):
    def __init__(self, texts, labels, tokenizer):
        self.encodings = tokenizer(
            texts, padding=True, truncation=True, max_length=512, return_tensors="pt"
        )
        self.labels = torch.tensor(labels, dtype=torch.float)

    def __getitem__(self, idx):
        return {
            'input_ids': self.encodings['input_ids'][idx],
            'attention_mask': self.encodings['attention_mask'][idx],
            'labels': self.labels[idx]
        }

    def __len__(self):
        return len(self.labels)

if __name__ == "__main__":
    MODEL_NAME = "bert-base-multilingual-cased"

    tokenizer = BertTokenizer.from_pretrained(MODEL_NAME)

    train_dataset = CustomDataset(X_train, y_train, tokenizer)
    test_dataset = CustomDataset(X_test, y_test, tokenizer)

    model = BertForSequenceClassification.from_pretrained(
        MODEL_NAME,
        num_labels=y_train.shape[1],
        problem_type="multi_label_classification"
    )

    training_args = TrainingArguments(
        output_dir='./multi_bert_results',
        per_device_train_batch_size=16,
        num_train_epochs=4,
        eval_strategy="epoch",
        save_strategy="epoch",
        logging_dir='./logs_multi_bert',
        logging_steps=50,
        load_best_model_at_end=True,
        metric_for_best_model="eval_loss",
        fp16=False,
        dataloader_num_workers=0 
    )

    trainer = Trainer(
        model=model,
        args=training_args,
        train_dataset=train_dataset,
        eval_dataset=test_dataset,
    )

    trainer.train()

    predictions = trainer.predict(test_dataset)
    y_pred_logits = predictions.predictions
    y_pred = (torch.sigmoid(torch.tensor(y_pred_logits)) > 0.5).int().numpy()

    print("Classification Report:")
    print(classification_report(y_test, y_pred, digits=4))

    print("\nДополнительные метрики:")
    print(f"Micro F1-score:     {f1_score(y_test, y_pred, average='micro'):.4f}")
    print(f"Macro F1-score:     {f1_score(y_test, y_pred, average='macro'):.4f}")
    print(f"Weighted F1-score:  {f1_score(y_test, y_pred, average='weighted'):.4f}")
    print(f"Hamming loss:       {hamming_loss(y_test, y_pred):.4f}")
    print(f"Subset accuracy:    {accuracy_score(y_test, y_pred):.4f} (все метки совпали)")



y_pred_prob = torch.sigmoid(torch.tensor(y_pred_logits)).numpy()

roc_auc = roc_auc_score(Y_TEST, y_pred_prob, average='macro')
print(f"ROC AUC (macro): {roc_auc:.4f}")


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-multilingual-cased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Classification Report:
              precision    recall  f1-score   support

           0     0.9517    0.9741    0.9628       425
           1     0.8571    0.7742    0.8136        31
           2     0.9012    0.8795    0.8902        83
           3     0.9398    0.9286    0.9341       168
           4     0.7000    0.6364    0.6667        88
           5     0.7727    0.6538    0.7083        52

   micro avg     0.9077    0.8937    0.9007       847
   macro avg     0.8538    0.8078    0.8293       847
weighted avg     0.9038    0.8937    0.8981       847
 samples avg     0.9091    0.9031    0.8933       847


Дополнительные метрики:
Micro F1-score:     0.9007
Macro F1-score:     0.8293
Weighted F1-score:  0.8981
Hamming loss:       0.0587
Subset accuracy:    0.7278 (все метки совпали)
ROC AUC (macro): 0.9408


c:\Users\josha\OneDrive\Documents\GitHub\4_intensiv_4\.venv\Lib\site-packages\sklearn\metrics\_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in samples with no true labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [ ]:

model.save_pretrained("./saved_model1")
tokenizer.save_pretrained("./saved_model1")

('./saved_model\\tokenizer_config.json',
 './saved_model\\special_tokens_map.json',
 './saved_model\\vocab.txt',
 './saved_model\\added_tokens.json')

RuRoBERTa 

In [42]:
tokenizer = RobertaTokenizer.from_pretrained("sberbank-ai/ruRoberta-large")

class CustomDataset(Dataset):
    def __init__(self, texts, labels, tokenizer):
        self.encodings = tokenizer(
            texts, padding=True, truncation=True, max_length=512, return_tensors="pt"
        )
        self.labels = torch.tensor(labels, dtype=torch.float)

    def __getitem__(self, idx):
        return {
            'input_ids': self.encodings['input_ids'][idx],
            'attention_mask': self.encodings['attention_mask'][idx],
            'labels': self.labels[idx]
        }

    def __len__(self):
        return len(self.labels)

if __name__ == "__main__":
    MODEL_NAME = "sberbank-ai/ruRoberta-large"


    tokenizer = RobertaTokenizer.from_pretrained(MODEL_NAME)

    train_dataset = CustomDataset(X_train, y_train, tokenizer)
    test_dataset = CustomDataset(X_test, y_test, tokenizer)

    model = RobertaForSequenceClassification.from_pretrained(
        MODEL_NAME,
        num_labels=y_train.shape[1],
        problem_type="multi_label_classification"
    )

    training_args = TrainingArguments(
        output_dir='./ruroberta_results',
        per_device_train_batch_size=8,
        num_train_epochs=4,
        eval_strategy="epoch",
        save_strategy="epoch",
        logging_dir='./logs_ruroberta',
        logging_steps=50,
        load_best_model_at_end=True,
        metric_for_best_model="eval_loss",
        fp16=False,
        dataloader_num_workers=0  
    )

    trainer = Trainer(
        model=model,
        args=training_args,
        train_dataset=train_dataset,
        eval_dataset=test_dataset,
    )

    trainer.train()

    predictions = trainer.predict(test_dataset)
    y_pred_logits = predictions.predictions
    y_pred = (torch.sigmoid(torch.tensor(y_pred_logits)) > 0.5).int().numpy()

    print("Classification Report:")
    print(classification_report(y_test, y_pred, digits=4))

    print("\nДополнительные метрики:")
    print(f"Micro F1-score:     {f1_score(y_test, y_pred, average='micro'):.4f}")
    print(f"Macro F1-score:     {f1_score(y_test, y_pred, average='macro'):.4f}")
    print(f"Weighted F1-score:  {f1_score(y_test, y_pred, average='weighted'):.4f}")
    print(f"Hamming loss:       {hamming_loss(y_test, y_pred):.4f}")
    print(f"Subset accuracy:    {accuracy_score(y_test, y_pred):.4f} (все метки совпали)")


    y_pred_prob = torch.sigmoid(torch.tensor(y_pred_logits)).numpy()

    roc_auc = roc_auc_score(Y_TEST, y_pred_prob, average='macro')
    print(f"ROC AUC (macro): {roc_auc:.4f}")

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at sberbank-ai/ruRoberta-large and are newly initialized: ['classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.bias', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


KeyboardInterrupt: 

In [ ]:
model.save_pretrained("./saved_model2")
tokenizer.save_pretrained("./saved_model2")